src/transform/src/threshold_elision.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Remove Threshold operators when we are certain no records have negative multiplicity.
+//!
+//! If we have Threshold(A - Subset(A)) and we believe that A has no negative multiplicities,
+//! then we can replace this with A - Subset(A).
+//!
+//! The Subset(X) notation means that the collection is a multiset subset of X:
+//! multiplicities of each record in Subset(X) are at most that of X.
+
+use std::collections::HashSet;
+
+use crate::TransformArgs;
+use mz_expr::visit::Visit;
+use mz_expr::MirRelationExpr;
+use mz_expr::{Id, LocalId};
+
+/// Remove Threshold operators that have no effect.
+#[derive(Debug)]
+pub struct ThresholdElision;
+
+impl crate::Transform for ThresholdElision {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.try_visit_mut_post(&mut |e| self.action(e))
+    }
+}
+
+impl ThresholdElision {
+    /// Remove Threshold operators with no effect.
+    pub fn action(&self, relation: &mut MirRelationExpr) -> Result<(), crate::TransformError> {
+        if let MirRelationExpr::Threshold { input } = relation {
+            // We look for the pattern `Union { base, Negate(Subset(base)) }`.
+            let mut should_replace = false;
+            if let MirRelationExpr::Union { base, inputs } = &mut **input {
+                if inputs.len() == 1 {
+                    if let MirRelationExpr::Negate { input } = &inputs[0] {
+                        let mut safe_lets = HashSet::default();
+                        if non_negative(base, &mut safe_lets) && subset_of(base, &*input) {
+                            should_replace = true;
+                        }
+                    }
+                }
+            }
+            if should_replace {
+                *relation = input.take_dangerous();
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Return true iff `relation` is believed to contain no negative multiplicities.
+///
+/// This relies on all `Get` bindings corresponding to collections without negative
+/// multiplicities. Local let bindings present in `safe_lets` are relied on to have
+/// no non-negative multiplicities.
+pub fn non_negative(relation: &MirRelationExpr, safe_lets: &mut HashSet<LocalId>) -> bool {
```

Comment:
This method needs to be rewritten so it is stack safe. 

*Side note*: In general, the value of this function for each sub-tree is an example of an attribute that can be derived bottom-up based on some context (the value derived for all local let bindings that are in scope). I can imagine other transformations that might benefit from this information, so it will be great if we can make this logic modular and reusable. I might use it as an example for a formalism I want to present at some point in the future.